clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -201,9 +215,23 @@ void run(long now) {
         }
 
         List<RecordBatch> expiredBatches = this.accumulator.abortExpiredBatches(this.requestTimeout, now);
-        // update sensors
-        for (RecordBatch expiredBatch : expiredBatches)
+
+        // Reset the PID if an expired batch has previously been sent to the broker. Aloso update the metrcis
+        // for expired batches.
+        boolean resetState = false;
+        for (RecordBatch expiredBatch : expiredBatches) {
+            if (transactionState != null && expiredBatch.inRetry()) {
+                resetState = true;
+            }
             this.sensors.recordErrors(expiredBatch.topicPartition.topic(), expiredBatch.recordCount);
+        }
+
+        if (resetState) {
+            transactionState.reset();
+            while (!transactionState.hasPid()) {
+                getPID(requestTimeout);
```

Comment:
I've gone with `Thread.sleep` after confirming with Jason that there is no better way in our code at present (ie. no wrapper or other abstraction for such a thing).